### PR TITLE
Duty Scheduler

### PR DESCRIPTION
**Pull request by olegshmuelov**
## 🎯 Refactoring of Duty Scheduler for SSV Node

This pull request introduces substantial refactoring of the Duty Scheduler in the SSV Project. The changes aim to enhance the current architecture by optimizing duty scheduling for upcoming epochs/periods 🚀, improving the execution speed of Attester and Sync Committee duties 🏃‍♂️, handling blockchain reorganizations effectively 🔄, and proactively scheduling duties for upcoming active validators 🔮.

The refactoring restructures the Duty Scheduler and Duty Fetcher, into one primary component that manage validator duties in the Ethereum 2.0 protocol.

**🔑 Key Improvements**:
- The Duty Scheduler now not only manages current epoch duties, but also plans and schedules duties for upcoming epochs and periods .
- The scheduler has been improved to handle re-organizations in blockchain state and to 
- Foresee and schedule duties for validators expected to become active in the beacon chain in the next epoch or sync committee period.
- The scheduler now efficiently handles (un)/scheduling duties in response to contract events

These enhancements aim to provide a more robust, efficient and proactive duty scheduling system, effectively handling upcoming duties and responding promptly to changes in the blockchain state and contract events. Further discussions and testing are required to ensure the efficiency and reliability of the new design 🧪.

### CODE REVIEW

File: operator/duties/attester.go
Comment (nkryuchkov) [inline operator/duties/attester.go:232]:
```suggestion
	}
	if currentSlot+1 == duty.Slot {
```

File: operator/duties/base_handler.go
Comment (nkryuchkov) [inline operator/duties/base_handler.go:66]: Should the map be locked with a mutex?

File: operator/duties/proposer.go
Comment (nkryuchkov) [inline operator/duties/proposer.go:60]: Let's add a case for `<-ctx.Done`

File: operator/node.go
Comment (nkryuchkov) [inline operator/node.go:159]: Let's not use `"github.com/pkg/errors"` as it's discontinued
```suggestion
		return fmt.Errorf("run duty scheduler: %w", err)
```

File: operator/duties/validatorregistration.go
Comment (nkryuchkov) [inline operator/duties/validatorregistration.go:18]: Do we need a mutex or https://github.com/cornelk/hashmap ?

File: operator/duties/validatorregistration.go
Comment (nkryuchkov) [inline operator/duties/validatorregistration.go:35]: Let's check for `<-ctx.Done()`

File: operator/duties/scheduler.go
Comment (nkryuchkov) [inline operator/duties/scheduler.go:292]: Can we have a test for this method to make sure that locking `waitMutex` here and in the goroutine doesn't lead to a deadlock?

File: operator/duties/attester.go
Comment (nkryuchkov) [inline operator/duties/attester.go:61]: Let's pass logger to NewAttesterHandler instead of passing it to each of its methods

File: operator/duties/base_handler.go
Comment (nkryuchkov) [inline operator/duties/base_handler.go:35]: Should this be atomic?

File: logging/fields/fields.go
Comment (nkryuchkov) [inline logging/fields/fields.go:68]: let's keep in sorted alphabetically

File: operator/duties/scheduler.go
Comment (nkryuchkov) [inline operator/duties/scheduler.go:101]:
```suggestion
		return fmt.Errorf("failed to subscribe to head events: %w", err)
```

File: operator/duties/attester.go
Comment (olegshmuelov) [reply operator/duties/attester.go]: done

File: operator/duties/proposer.go
Comment (olegshmuelov) [reply operator/duties/proposer.go:63]: done

File: operator/duties/validatorregistration.go
Comment (olegshmuelov) [reply operator/duties/validatorregistration.go]: done

File: logging/fields/fields.go
Comment (olegshmuelov) [reply logging/fields/fields.go]: done

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go]: done

File: operator/duties/base_handler.go
Comment (olegshmuelov) [reply operator/duties/base_handler.go:34]: no, the code designed to not write in parallel

File: beacon/goclient/sync_committee_contribution.go
Comment (y0sher) [inline beacon/goclient/sync_committee_contribution.go:96]: I don't see where does a valid block comes in this func. its only waitOneThird

File: go.mod
Comment (y0sher) [inline go.mod:154]: very nice lib, good find.

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:68]: this could be just 
```
if h.shouldFetchNexEpoch(h.network.Beacon.EstimatedCurrentSlot()) {
		h.fetchNextEpoch = true
	}
```
if you would set `h.fetchCurrentEpoch` and `h.fetchFirst` to true in `NewAttesterHandler`

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:89]: shouldn't we check if indicies changed before execution? also shouldn't we check it even if `fetchFirst` is true?

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:90]: it is possible that  we fetched stuff that should be executed on this slot? in other words should we run execution again after fetching?

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:102]: why not set a var 
```suggestion

                        slotsPerEpoch := network.Beacon.SlotsPerEpoch()
			// Get next epoch's attester duties, but wait until half-way through the epoch
			// This allows us to set them up at a time when the beacon node should be less busy.
			if uint64(slot)%slotsPerEpoch == slotsPerEpoch/2-2 {
				h.fetchNextEpoch = true
			}

			// last slot of epoch
			if uint64(slot)%slotsPerEpoch == slotsPerEpoch-1 {
				h.duties.Reset(currentEpoch)
			}
```

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:93]: rephrasing suggestion
```suggestion
			// If we're half-way through the epoch, get next epoch's attester duties.
```

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:189]: name is kinda funny since you log in this func and not just prepare

File: operator/duties/validatorregistration.go
Comment (olegshmuelov) [reply operator/duties/validatorregistration.go:18]: no need, there no parallel access to the map

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go]: the unit tests covering it

File: beacon/goclient/sync_committee_contribution.go
Comment (olegshmuelov) [reply beacon/goclient/sync_committee_contribution.go:96]: the purpose of this func to release the waiting 1/3 of the slot time on timeout or on valid block arrival

**Review by moshe-blox — CHANGES_REQUESTED** [review]

File: ekm/eth_key_manager_signer.go
Comment (olegshmuelov) [reply ekm/eth_key_manager_signer.go:69]: GetBeaconNetwork
should return the spec network

GetNetwork
should return the SSV network

all this area should be refactored starting from the spec

File: operator/duties/attester.go
Comment (olegshmuelov) [reply operator/duties/attester.go:91]: no, the new fetched duties should be executed only on the next slot

File: ekm/eth_key_manager_signer.go
Comment (olegshmuelov) [reply ekm/eth_key_manager_signer.go:69]: see https://github.com/bloxapp/ssv/pull/1044#discussion_r1269319253

File: operator/duties/proposer.go
Comment (y0sher) [inline operator/duties/proposer.go:133]: I think fetchDuties should only fetch duties, return them and the calling function should decide what to do with it (ex: add to the map, log etc..). otherwise this function name should be different.

Comment (olegshmuelov) [reply operator/duties/proposer.go]: we want to have one place to have it all (the fetch, the log the subscribe and so on)
we can rename it to fetchAndProcessDuties

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go]: when you call this function you usually already know the current slot, so passing it in instead of calling `h.network.Beacon.EstimatedCurrentSlot()` can reduce times you call it and reduce the dependency in this function.

Comment (olegshmuelov) [reply operator/duties/attester.go]: We explicitly estimate the current slot at the exact time of the function call to ensure our operation is based on the most accurate and recent data

File: beacon/goclient/sync_committee_contribution.go
Comment (y0sher) [reply beacon/goclient/sync_committee_contribution.go:96]: Again I can't see how this function is waiting to the block arrival. you might use it in another place to do so, so this function name should include only what this function does.

File: operator/duties/attester.go
Comment (y0sher) [inline operator/duties/attester.go:174]: this might be just code-style but I prefer less nested code. feel free to dismiss..
```suggestion
	slotMap, ok := h.duties.m[epoch];
	if !ok
	   return
	}
	duties, ok := slotMap[slot]
	if !ok {
	    return
	}
	toExecute := make([]*spectypes.Duty, 0, len(duties)*2)
	for _, d := range duties {
		if h.shouldExecute(d) {
			toExecute = append(toExecute, h.toSpecDuty(d, spectypes.BNRoleAttester))
			toExecute = append(toExecute, h.toSpecDuty(d, spectypes.BNRoleAggregator))
		}
	}
	h.executeDuties(h.logger, toExecute)
```

File: beacon/goclient/sync_committee_contribution.go
Comment (olegshmuelov) [reply beacon/goclient/sync_committee_contribution.go]: was referring to diff func :)
np renamed it to `waitForOneThirdSlotDuration`

File: operator/duties/scheduler.go
Comment (y0sher) [reply operator/duties/scheduler.go:181]: two approaches can be taken, 
- in, and subscribers channels should be buffered
- add a default clause to the select.

I guess the first one is the best because the default will mean we will skip events and that will cause issues.

Comment (y0sher) [reply operator/duties/scheduler.go:181]: we could also fall into default and retry only the subscribers that were blocking, still I think a small buffer will be ok since none of these event should happen a lot.

File: operator/duties/scheduler.go
Comment (y0sher) [inline operator/duties/scheduler.go:206]: maybe this should be a const?

File: operator/duties/scheduler.go
Comment (y0sher) [inline operator/duties/scheduler.go:336]: aren't we waiting only for the block here?

File: operator/validator/controller.go
Comment (y0sher) [inline operator/validator/controller.go:161]: this could be an empty struct `indicesChange chan struct{}` since we never use the boolean?

File: operator/validator/event_handler.go
Comment (y0sher) [inline operator/validator/event_handler.go:1]: we don't want to signal removed validators as well?

File: operator/validator/event_handler.go
Comment (y0sher) [inline operator/validator/event_handler.go:395]: what if last one in range wasn't started, but 2 befores did. we still need to shoot an event to the channel. this won't work. last one is what matters in this code. I suggest maybe stop changing started after one has returned true.

**Review by y0sher — CHANGES_REQUESTED** [review]
Please go through my comments. generally good work, I love the thorough testing 🥇

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go:206]: do you mean move the `3` to be a const?

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go:336]: no, we are waiting for the block from head event or for the one third of the slot release from scheduler slot ticker

File: operator/validator/controller.go
Comment (olegshmuelov) [reply operator/validator/controller.go:161]: see https://github.com/bloxapp/ssv/pull/1044#discussion_r1273232868

File: operator/validator/event_handler.go
Comment (olegshmuelov) [reply operator/validator/event_handler.go:395]: good catch

File: operator/duties/scheduler.go
Comment (moshe-blox) [reply operator/duties/scheduler.go:181]: the problem with buffering is that if you were stuck for say 12 seconds (the max timeout), then after 12 sec after the slot has already advanced, you resume reading from slotTicker, but then you get the previous slot from the ticker bc it was buffered

the default clause may work, we have this prototype:
```go
func fanOut[T any](ctx context.Context, in <-chan T, subscribers []chan<- T) {
    for {
        select {
        case <-ctx.Done():
            return
        case item := <-in:
            outbox := make([]chan<- T, len(subscribers))
            copy(outbox, subscribers)

            for len(outbox) > 0 {
                for i, subscriber := range outbox {
                    // Fan out the message to all subscribers.
                    select {
                    case <-ctx.Done():
                        return
                    case subscriber <- item:
                        // Sent -> Remove the subscriber from the outbox.
                        outbox = append(outbox[:i], outbox[i+1:]...)
                    default:
                    }
                }
            }
        }
    }
}
```

however bc it gets a bit complex, we're considering just using Prysm's event.Feed which does a similar thing already

File: operator/duties/scheduler.go
Comment (moshe-blox) [reply operator/duties/scheduler.go:206]: @y0sher @olegshmuelov 
this is called `INTERVALS_PER_SLOT` in the Beacon spec

https://github.com/prysmaticlabs/prysm/blob/2674e204cab117b1e902c78d72d8b6faaec7e6a0/config/params/mainnet_config.go#L119

Comment (olegshmuelov) [reply operator/duties/scheduler.go]: adjusted in https://github.com/bloxapp/ssv/pull/1044/commits/4c8d7399dc6536a36faba4c4e20aa4db22adf148

**Review by y0sher — APPROVED** [review]
Glad to see this new design and the added testing. good job!

**Review by moshe-blox — APPROVED** [review]

**Comment by andrew-blox** [comment]
Manual and Unit test plans can be found here:
https://docs.google.com/spreadsheets/d/15dB_kIRkc43cckAGyttlKTnJKFJtFTT8HdUHIjcUL_I/edit#gid=1806251730

**Review by andrew-blox — APPROVED** [review]

File: operator/duties/scheduler.go
Comment (y0sher) [inline operator/duties/scheduler.go:136]: what if an error is returned? subscribed is still called and we act like we have head events. I think there's no value in this code being inside a goroutine, it just adds complexity.

File: operator/validator/controller_test.go
Comment (y0sher) [inline operator/validator/controller_test.go:103]: why not use current epoch?

**Review by y0sher — CHANGES_REQUESTED** [review]
last small comment and lets merge

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go]: see [only call subscribed if there was no error](https://github.com/bloxapp/ssv/pull/1044/commits/0ed04167602bfd2f7592b0ada5fd6357aff7b35a)

File: operator/validator/controller_test.go
Comment (olegshmuelov) [reply operator/validator/controller_test.go:103]: I chose to use the hardcoded value `phase0.Epoch(100)` for the following reasons:

1. **Ensuring a Known State**: Tests should ideally be deterministic, meaning they always produce the same results given the same input. Using a hardcoded value helps to ensure this, as it prevents the test's behavior from changing based on the state of an external variable like "currentEpoch".

2. **Readability**: By using a specific value, it can be clearer to someone reading the test why a specific scenario is being tested.

3. **Isolation**:  If "currentEpoch" is set elsewhere in the code, changing its value could potentially break this test, even if there's nothing wrong with the code being tested. Using a hardcoded value avoids this problem.

Hope this clarifies the decision :)

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go]: There is a convention in golang to use the Start function to initiate an operation in a non-blocking manner, often launching goroutines to do the work in the background.

the subscription to head events is running in a goroutine because it needs to be non-blocking. If the subscription were moved out of the goroutine, it could potentially block the rest of the Start function, which goes against the convention.

Comment (y0sher) [reply operator/duties/scheduler.go]: This convention adheres to long actions. Subscribing to headevents should be a short action, and since we don't want to continue if we can't subscribe, it's better to do it in a synchronous matter. After [only call subscribed if there was no error](https://github.com/bloxapp/ssv/pull/1044/commits/0ed04167602bfd2f7592b0ada5fd6357aff7b35a), the situation is now worse, if there was an error all our handler go routines will hang waiting to the subscribe context to be done but it won't event be done, the handler will never run.
When you write a start function, it's best to run in the main start go routine stuff that are mandatory for the function to run, then return error if they fail, even if it blocks for a short time. after that you can start any long running tasks in go routine and unblock.

Comment (y0sher) [reply operator/duties/scheduler.go]: See this exmaple for a start function

```golang

func Start(input string) error {

    if input != GoodInput {
        return ErrBadInput
    }
    if err := CheckRequiredService(); err != nil {
        return err
    }

    go ServiceEventLoop()

    return nil

}

File: operator/validator/controller_test.go
Comment (y0sher) [reply operator/validator/controller_test.go:103]: Honestly since you defined the currentEpoch variable in the test and your testing against it, I think its better to use it. In the future if we'll want to change the value you'll need to edit many places in stead of one.

1. this is incorrect, currectEpoch is an internal test variable. its basically like hardcoded
2. its easier to see differences when using this variable
3. this is incorrect, its set a few lines above and not "elsewhere in the code". 
your reusing the same hardcoded value when you defined a var for this purpose.

File: operator/duties/scheduler.go
Comment (olegshmuelov) [reply operator/duties/scheduler.go]: > the situation is now worse, if there was an error all our handler go routines will hang waiting to the subscribe context to be done but it won't ever be done, the handler will never run.

if there is an error subscribing 
`s.pool = pool.New().WithContext(ctx).WithCancelOnError()`
should cancel the context on error 

and the `Wait`
```
if err := n.dutyScheduler.Wait(); err != nil {
  logger.Fatal("duty scheduler exited with error", zap.Error(err))
}
```
should catch it

**Review by y0sher — APPROVED** [review]
Lets merge this!

**Review by moshe-blox — APPROVED** [review]

**Pull request merged**
